src/library/scala/collection/immutable/TreeSet.scala

```diff
@@ -25,11 +25,65 @@ import mutable.{ Builder, SetBuilder }
 object TreeSet extends ImmutableSortedSetFactory[TreeSet] {
   implicit def implicitBuilder[A](implicit ordering: Ordering[A]): Builder[A, TreeSet[A]] = newBuilder[A](ordering)
   override def newBuilder[A](implicit ordering: Ordering[A]): Builder[A, TreeSet[A]] =
-    new SetBuilder(empty[A](ordering))
+    new TreeSetBuilder
 
   /** The empty set of this type
    */
   def empty[A](implicit ordering: Ordering[A]) = new TreeSet[A]
+  private class TreeSetBuilder[A](implicit val ordering: Ordering[A]) extends Builder[A, TreeSet[A]] {
+    type Tree = RB.Tree[A, Unit]
+    private [this] var tree:Tree = null
+    override def +=(elem: A): TreeSetBuilder.this.type = {
+      tree = RB.update(tree, elem, (), overwrite = false)
+      this
+    }
+
+    private object adder extends Function1[A, Unit] {
```

Comment:
Of course, you could make `TreeSetBuilder` extend `A => Unit` and save the object overhead, but that's too clever by at least half